### PR TITLE
Add support for multi-user key store providers

### PR DESCRIPTION
**Pull request by karinazhou**
The following changes are made:
- Added connection instance-level column encryption key store providers
- Added statement instance-level column encryption key store providers
- Added cache mechanism to Azure Key Vault provider 
- Disabled the CEK caches on system providers and global providers
- Updated the key finding logic 
- Updated related APIs to support instance-level providers
- Added tests

Initial draft: https://github.com/David-Engel/mssql-jdbc/tree/multitenant

### CODE REVIEW

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 1]
```diff
@@ -505,16 +505,19 @@ protected static void createScaleTable(String tableName, String table[][], Strin
...
+                .getConnection(connectionString + ";sendTimeAsDateTime=false", AEInfo);
                 SQLServerStatement stmt = (SQLServerStatement) con.createStatement()) {
             String sql = " if not exists (SELECT name from sys.column_master_keys where name='" + cmkName + "')"
                     + " begin" + " CREATE COLUMN MASTER KEY " + cmkName + " WITH (KEY_STORE_PROVIDER_NAME = '"
                     + keyStoreName + "', KEY_PATH = '" + keyPath + "'"
                     + (TestUtils.isAEv2(con) ? ",ENCLAVE_COMPUTATIONS (SIGNATURE = " + signature + ")) end" : ") end");
             stmt.execute(sql);
+        } catch (Exception ex) {
+            // kz debug
+            ex.printStackTrace();
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java:520]: check for expected error msg and fail otherwise.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 2]
```diff
@@ -0,0 +1,643 @@
...
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo)) {           
+            try (SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sql,
+                SQLServerStatementColumnEncryptionSetting.Enabled)) {
+                pstmt.setInt(1, customerId);
+                ResultSet rs = pstmt.executeQuery(); 
+                fail(TestResource.getResource("R_expectedExceptionNotThrown"));
+            } catch (Exception ex) {
+                // kz debug
+                System.out.println(ex.getMessage());
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:353]: multiple places - remove debug

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 3]
```diff
@@ -0,0 +1,643 @@
...
+@Tag(Constants.xSQLv12)
+@Tag(Constants.xAzureSQLDW)
+@Tag(Constants.xAzureSQLDB)
+public class MultiUserAKVTest extends AESetup {
+
+    private static Map<String, SQLServerColumnEncryptionKeyStoreProvider> requiredKeyStoreProvider = new HashMap<>();
+    private static Map<String, SQLServerColumnEncryptionKeyStoreProvider> notRequiredKeyStoreProvider = new HashMap<>();
+    
+    private static final String notRequiredProviderName = "DummyProvider2";
+    private static final String requiredProviderName = "DUMMY_PROVIDER";
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:65]: nitpick - can we be more consistent in use of capitalization in the names

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 4]
```diff
@@ -0,0 +1,643 @@
...
+            byte[] plaintextKey = { 1, 2, 3 };
+            byte[] encryptedKey = providerInGlobalCache.encryptColumnEncryptionKey(keyIDs[0], Constants.CEK_ALGORITHM, plaintextKey);
+
+            providerInGlobalCache.decryptColumnEncryptionKey(keyIDs[0], Constants.CEK_ALGORITHM, encryptedKey);
+            assertEquals(0, getCacheSize(cekCacheSizeGetterName, providerInGlobalCache));
+        } catch (SQLException e) {
+            fail(e.getMessage());
+        } finally {
+            // clean up
+            SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:239]: do we need this? it's done in AfterAll

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 5]
```diff
@@ -0,0 +1,643 @@
...
+        
+        String sql = "SELECT CustomerId, CustomerName FROM " + customProviderTableName + " WHERE CustomerId = ?";
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo);
+                SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sql,
+                    SQLServerStatementColumnEncryptionSetting.Enabled)) {            
+            
+            pstmt.setInt(1, customerId);
+            
+            ResultSet rs = pstmt.executeQuery();
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:273]: leak - use try with resources (same in line 288)

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java [hunk 6]
```diff
@@ -96,17 +98,80 @@ static boolean compareBytes(byte[] buffer1, byte[] buffer2, int buffer2Index, in
...
+    static boolean shouldUseInstanceLevelProviderFlow(String keyStoreName, SQLServerConnection connection, SQLServerStatement statement) {
+        return !keyStoreName.equalsIgnoreCase(WINDOWS_KEY_STORE_NAME) 
+            && (connection.hasConnectionColumnEncryptionKeyStoreProvidersRegistered() || (null != statement && statement.hasColumnEncryptionKeyStoreProvidersRegistered()));
+        
+        // kz for debug
+        // boolean result1 = !keyStoreName.equalsIgnoreCase(WINDOWS_KEY_STORE_NAME);
+        // boolean result2 = connection.hasConnectionColumnEncryptionKeyStoreProvidersRegistered();
+        // boolean result3 = (null != statement && statement.hasColumnEncryptionKeyStoreProvidersRegistered());
+
+        // return result1 && (result2 || result3);
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java:121]:
```suggestion
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSymmetricKeyCache.java [hunk 7]
```diff
@@ -137,15 +108,12 @@ SQLServerSymmetricKey getKey(EncryptionKeyInfo keyInfo, SQLServerConnection conn
...
-                    cache.putIfAbsent(keyLookupValue, encryptionKey);
-                    if (aeLogger.isLoggable(java.util.logging.Level.FINE)) {
-                        aeLogger.fine("Adding encryption key to cache...");
-                    }
-                    scheduler.schedule(new CacheClear(keyLookupValue), columnEncryptionKeyCacheTtl, SECONDS);
+                    cache.setCacheTtl(columnEncryptionKeyCacheTtl);
+                    cache.put(keyLookupValue, encryptionKey);
                 }
-            } else {
+            }else {
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSymmetricKeyCache.java:114]:
```suggestion
            } else {
```

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSymmetricKeyCache.java [hunk 8]
```diff
@@ -137,15 +108,12 @@ SQLServerSymmetricKey getKey(EncryptionKeyInfo keyInfo, SQLServerConnection conn
...
                 encryptionKey = cache.get(keyLookupValue);
-            }
+            }
```
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSymmetricKeyCache.java:116]:
```suggestion
            }
```

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java [hunk 9]
```diff
@@ -61,7 +61,7 @@
     static String cekJks = Constants.CEK_NAME + "_JKS";
     static String cekWin = Constants.CEK_NAME + "_WIN";
     static String cekAkv = Constants.CEK_NAME + "_AKV";
-
+
```
Comment (David-Engel) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java:64]:
```suggestion

```

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 10]
```diff
@@ -0,0 +1,643 @@
...
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providerMap = new HashMap<String, SQLServerColumnEncryptionKeyStoreProvider>();
+        providerMap.put(Constants.AZURE_KEY_VAULT_NAME, provider);
+        SQLServerConnection.registerColumnEncryptionKeyStoreProviders(providerMap);
+
+        try (SQLServerConnection con = (SQLServerConnection) PrepUtil.getConnection(AETestConnectionString, AEInfo)) {
+            Field globalCustomCacheField = SQLServerConnection.class.getDeclaredField("globalCustomColumnEncryptionKeyStoreProviders");
+            globalCustomCacheField.setAccessible(true);
+
+            Map<String, SQLServerColumnEncryptionKeyStoreProvider> globalCacheFieldValue = (Map<String, SQLServerColumnEncryptionKeyStoreProvider>) globalCustomCacheField.get(con);
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:225]: Eclipse warning:
Multiple markers at this line
	- Type safety: Unchecked cast from Object to Map<String,
	 SQLServerColumnEncryptionKeyStoreProvider>
	- Type safety: Unchecked cast from Object to Map<String,
	 SQLServerColumnEncryptionKeyStoreProvider>

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 11]
```diff
@@ -0,0 +1,643 @@
...
+            }
+            
+            // Register invalid key store provider on statement level. This will overwrite the previous one.
+            SQLServerColumnEncryptionAzureKeyVaultProvider providerWithBadCred = new SQLServerColumnEncryptionAzureKeyVaultProvider("badApplicationID", "badApplicationKey");
+            providerMap.put(Constants.AZURE_KEY_VAULT_NAME, providerWithBadCred);
+            pstmt.registerColumnEncryptionKeyStoreProvidersOnStatement(providerMap);
+
+            // The following query should fail due to an empty cek cache and invalid credentials
+            try (ResultSet rs3 = pstmt.executeQuery()) {
+                int numberOfColumns = rs3.getMetaData().getColumnCount();
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:302]: unused variable warning in eclipse (also line 349, 363, 378, 391, 437, 446, 456, 466, 479

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 12]
```diff
@@ -0,0 +1,643 @@
...
+    private void insertData(String tableName, int customId, String customName) {
+        String sqlQuery = "INSERT INTO " + tableName + " VALUES ( ?, ? )"; 
+
+        try (SQLServerConnection con = PrepUtil.getConnection(AETestConnectionString, AEInfo);
+                SQLServerPreparedStatement pstmt = (SQLServerPreparedStatement) TestUtils.getPreparedStmt(con, sqlQuery,
+                    SQLServerStatementColumnEncryptionSetting.Enabled)) { 
+            pstmt.setInt(1, customId);
+            pstmt.setString(2, customName);
+            pstmt.executeUpdate();
+            System.out.println("1 Record inserted successfully.");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:505]: remove debug line

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 13]
```diff
@@ -0,0 +1,643 @@
...
+        try (SQLServerConnection con = PrepUtil.getConnection(connString, AEInfo);
+                SQLServerStatement stmt = (SQLServerStatement) con.createStatement()) { 
+            TestUtils.dropTableIfExists(tableName, stmt);
+            stmt.execute(sqlQuery);                              
+        } catch(SQLException e) {
+            fail(e.getMessage());
+        }        
+    }
+
+    private void dropObject(String connString, String objectType, String objectName) {
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:526]: add this to TestUtils? (there is a dropObjectIfExists)

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 14]
```diff
@@ -0,0 +1,643 @@
...
+
+        MessageDigest md = MessageDigest.getInstance("SHA-256");
+        md.update(Constants.AZURE_KEY_VAULT_NAME.toLowerCase().getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        md.update(masterKeyPath.toLowerCase().getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        
+        if (allowEnclaveComputations) {
+            md.update("true".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        } else {
+            md.update("false".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        }
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:574]: md.update(Boolean.toString(allowEnclaveComputations), .getBytes(java.nio.charset.StandardCharsets.UTF_16LE));

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 15]
```diff
@@ -0,0 +1,643 @@
...
+        return signature;
+    }
+    
+    private int getCacheSize(String methodName, SQLServerColumnEncryptionAzureKeyVaultProvider provider) throws Exception {
+        Method method = provider.getClass().getDeclaredMethod(methodName);
+        method.setAccessible(true);
+        
+        int count = (int)method.invoke(provider);
+
+        return count;
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:601]: return (int)method.invoke(provider);

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 16]
```diff
@@ -0,0 +1,643 @@
...
+    private int getCacheSize(String methodName, SQLServerColumnEncryptionAzureKeyVaultProvider provider) throws Exception {
+        Method method = provider.getClass().getDeclaredMethod(methodName);
+        method.setAccessible(true);
+        
+        int count = (int)method.invoke(provider);
+
+        return count;     
+    }
+
+    private SQLServerColumnEncryptionAzureKeyVaultProvider createAKVProvider() throws Exception {
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:604]: why is this necessary? It's already done in setup() in AbstractTest?

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 17]
```diff
@@ -0,0 +1,643 @@
...
+                if (null != file) {
+                    file.delete();
+                }
+            }
+        }
+
+        return azureKeyVaultProvider;
+    }
+
+    private static String byteToHexDisplayString(byte[] b) {
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:630]: can we use bytesToHexString in TestUtil? if not, this should go there

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 18]
```diff
@@ -0,0 +1,643 @@
...
+
+            // The following query should fail due to an empty cek cache and invalid credentials
+            try (ResultSet rs3 = pstmt.executeQuery()) {
+                int numberOfColumns = rs3.getMetaData().getColumnCount(); 
+                while (rs3.next()) {
+                    int intValue = rs3.getInt(1);
+                    String strValue = rs3.getString(2);
+                    assertTrue((customerId == intValue) && strValue.equalsIgnoreCase(customerName));
+                }               
+                fail("Expected SQLServerException is not caught.");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:308]: use R_expectedExceptionNotThrown

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 19]
```diff
@@ -0,0 +1,643 @@
...
+        
+        if (allowEnclaveComputations) {
+            md.update("true".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        } else {
+            md.update("false".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        }
+
+        byte[] dataToSign = md.digest();
+        if (null == dataToSign) {
+            fail("data to sign is null or empty.");
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:578]: this isn't defined in TestResource. Can we have just a generic {0} is null or empty error for all these cases instead of adding each separately? (same for line 589)

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java (hunk 1 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/AESetup.java:520]: This is for debugging only. Will remove it when doing clean up.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 4 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:239]: This is to avoid conflict when running the following tests in the same file.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 13 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:526]: There are some functions like dropCEK / dropCMK from package `com.microsoft.sqlserver.jdbc.AlwaysEncrypted`. I have moved this to AESetup.java.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 16 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:604]: This is to create a new AKV provider instance that will have a fresh cache associated with it. I am trying to avoid the same akvProvider instance shared by other existing tests.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 17 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:630]: The return values are slightly different. I have moved it to TestUtils.java

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 19 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:578]: Added generic message template.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java [hunk 20]
```diff
@@ -25,6 +26,8 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.logging.Level;
 
+import org.spongycastle.util.Arrays;
```
[the remark this answers is not in the document]
Comment (karinazhou) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerColumnEncryptionAzureKeyVaultProvider.java:29]: It's not used. Removed.

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java [hunk 21]
```diff
@@ -178,11 +178,17 @@ default ResultSet executeSDPEv1(PreparedStatement stmt, String userSql,
...
             ArrayList<byte[]> enclaveRequestedCEKs) throws SQLException {
         Map<Integer, CekTableEntry> cekList = new HashMap<>();
         CekTableEntry cekEntry = null;
         boolean isRequestedByEnclave = false;
+        SQLServerStatement statement = (SQLServerStatement) ((SQLServerPreparedStatement) stmt);
+
+        if (null != sqlServerStatement && sqlServerStatement.hasColumnEncryptionKeyStoreProvidersRegistered()) {
+            statement.registerColumnEncryptionKeyStoreProvidersOnStatement(sqlServerStatement.statementColumnEncryptionKeyStoreProviders);
+        }
+
```
[the remark this answers is not in the document]
Comment (karinazhou) [reply src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerEnclaveProvider.java:191]: stmt is java.sql.preparedStatement that doesn't have the registered providers we added.

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 3 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:66]: Changed

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 10 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:231]: Suppressed

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 11 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java]: removed

File: src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+package com.microsoft.sqlserver.jdbc;
+
+import java.time.Duration;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+final class SimpleTtlCache<K,V> {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java:15]: add javadocs describing the class and how it's used

File: src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    V put(K key, V value) {
+        V previousValue = null;
+        long cacheTtlInSeconds = cacheTtl.getSeconds();
+        
+        if (0 < cacheTtlInSeconds) {
+            previousValue = cache.put(key, value);
+            if (simpleCacheLogger.isLoggable(java.util.logging.Level.FINE)) {
+                simpleCacheLogger.fine("Adding encryption key to cache...");
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java:100]: this is for debugging purpose right? maybe finest?

Comment (karinazhou) [reply src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java:164]: changed

File: src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java (hunk 22 above)
Comment (karinazhou) [reply src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java:22]: added

File: src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+ */
+@RunWith(JUnitPlatform.class)
+public class RegisterKeyStoreProviderTest extends AbstractTest {
+
+    private static final String dummyProviderName1 = "DummyProvider1";
+    private static final String dummyProviderName2 = "DummyProvider2";
+    private static final String dummyProviderName3 = "DummyProvider3";
+    private static final String dummyProviderName4 = "DummyProvider4";
+
+    private static final String WINDOWS_KEY_STORE_NAME = "MSSQL_CERTIFICATE_STORE";
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java:34]: Constants.WINDOWS_KEY_STORE_NAME

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java [hunk 25]
```diff
@@ -0,0 +1,612 @@
...
+            .escapeSingleQuotes(AbstractSQLGenerator.escapeIdentifier(RandomUtil.getIdentifier("JDBCCustomProvider")));
+
+    private static final String dummyProviderTableName = TestUtils
+    .escapeSingleQuotes(AbstractSQLGenerator.escapeIdentifier(RandomUtil.getIdentifier("JDBCDummyProvider")));
+    
+    private static String cmkDummy = Constants.CMK_NAME + "_DUMMY";  // For dummyKeyStoreProvider
+	private static String cekDummy = Constants.CEK_NAME + "_DUMMY";  // For dummyKeyStoreProvider
+
+    private static boolean isMasterKeyPathSetup = false;
+    private static final char[] hexChars = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java:81]: not used

File: src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java [hunk 26]
```diff
@@ -0,0 +1,196 @@
...
+        Map<String, SQLServerColumnEncryptionKeyStoreProvider> providers = new HashMap<>();
+        providers.put("DUMMY_PROVIDER", new DummyKeyStoreProvider());
+
+        SQLServerConnection.unregisterColumnEncryptionKeyStoreProviders();
+
+        try {
+            SQLServerConnection.registerColumnEncryptionKeyStoreProviders(providers);
+        }
+        catch (Exception e) {
+            fail(TestResource.getResource("R_unexpectedException"));
```
Comment (lilgreenbird) [inline src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java:102]: can you append e.getMessage() to this like in other places? That makes it easier to debug failures. (multiple places)

File: src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+                    logger.fine("Removed key from cache...");
+                }
+            }
+        }
+    }
+
+    static final private java.util.logging.Logger simpleCacheLogger = java.util.logging.Logger.getLogger("com.microsoft.sqlserver.jdbc.SimpleTtlCache");
+
+    private final ConcurrentHashMap<K, V> cache;
+    private Duration cacheTtl = Duration.ofHours(2);  // The default time-to-live is set to 2 hours.
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java:57]: can we define a constant for this at the top, makes it easier to see the hard code value

**Review by v-zhangw — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java (hunk 27 above)
Comment (karinazhou) [reply src/main/java/com/microsoft/sqlserver/jdbc/SimpleTtlCache.java]: removed

File: src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java (hunk 26 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java]: changed

File: src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java (hunk 25 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java]: removed

File: src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java (hunk 24 above)
Comment (karinazhou) [reply src/test/java/com/microsoft/sqlserver/jdbc/RegisterKeyStoreProviderTest.java]: changed

**Review by lilgreenbird — APPROVED** [review]
